### PR TITLE
Alerting: Support for optimistic locking for alert rules

### PR DESCRIPTION
**Pull request by yuri-tceretian**
**What this PR does / why we need it**:
Currently, in alerting API code we read alert rules in a transaction but nothing stops two concurrent transactions to try to access the same record and change it concurrently. 
When that happens the transaction that changed the records las - wins. Until now this was ok behavior. However, considering https://github.com/grafana/grafana/issues/48353 where issues need to be processed as a group, data consistency across the entire group becomes very important. 
When a user updates a group of rules the API reads all rules from the database and all rules of groups affected by the change (for example, moving from one group to another). Therefore, all changes in those groups should be consistent within a transaction. 

~This PR updates data store methods that are used to read rules from the database to support a new parameter `ForUpdate` that tells XORM that the select query should be followed with the command `FOR UPDATE`. This will make MySQL and Postgres databases put locks on the rows that are selected during the update flow.~

This PR updates the rule store method `UpdateAlertRules` to update records in the database using optimistic lock by field Version. This field already exists and is properly maintained. XORM provides support for optimistic locking out of box https://xorm.io/docs/chapter-06/1.lock/. 

**Which issue(s) this PR fixes**:
Related: https://github.com/grafana/grafana/issues/48353 https://github.com/grafana/grafana/pull/50051

**Special notes for your reviewer**:
This change has a side-effect. It makes it impossible usage of XORM's Update method that changes multiple records (see commit [9360261](https://github.com/grafana/grafana/pull/50274/commits/936026119b729a0770e562a96920464348cf1659)) because XORM automatically adds filter by version which makes the update command a noop. Please ignore the commit as this code is being removed in https://github.com/grafana/grafana/pull/50295 and I fixed it here just to make tests pass.

### CODE REVIEW

**Comment by yuri-tceretian** [comment]
Blocked by changes in https://github.com/grafana/grafana/pull/50295

File: pkg/services/ngalert/store/alert_rule.go [hunk 1]
```diff
@@ -322,15 +326,13 @@ func (st DBstore) GetRuleGroupInterval(ctx context.Context, orgID int64, namespa
 
 func (st DBstore) UpdateRuleGroup(ctx context.Context, orgID int64, namespaceUID string, ruleGroup string, interval int64) error {
 	return st.SQLStore.WithDbSession(ctx, func(sess *sqlstore.DBSession) error {
-		_, err := sess.Update(
-			ngmodels.AlertRule{IntervalSeconds: interval},
-			ngmodels.AlertRule{OrgID: orgID, RuleGroup: ruleGroup, NamespaceUID: namespaceUID},
-		)
+		updated := TimeNow()
+		_, err := sess.Query("UPDATE alert_rule SET interval_seconds = ?, updated = ?, version = version + 1 WHERE org_id = ? AND rule_group = ? AND namespace_uid = ?", interval, updated, orgID, ruleGroup, namespaceUID)
 		return err
```
Comment (yuri-tceretian) [inline pkg/services/ngalert/store/alert_rule.go:331]: This change is temporary to make tests pass. This method is removed in another PR.

**Review by JohnnyQQQQ — APPROVED** [review]
LGTM

File: pkg/services/ngalert/store/alert_rule.go (hunk 1 above)
Comment (JohnnyQQQQ) [reply pkg/services/ngalert/store/alert_rule.go:331]: PR: https://github.com/grafana/grafana/pull/50696

**Pull request merged**